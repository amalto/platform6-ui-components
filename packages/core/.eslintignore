dist/
loader/

src/components.d.ts