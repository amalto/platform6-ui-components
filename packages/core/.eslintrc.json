{
  "extends": ["../../.eslintrc.json", "plugin:@stencil/recommended", "plugin:jest/recommended", "plugin:jsx-a11y/recommended"],
  "parserOptions": {
    "project": ["./tsconfig.json", "./tsconfig.lint.json"],
    "ecmaFeatures": {
      "jsx": true // Allows for the parsing of JSX
    }
  },
  "plugins": ["jest"],

  "rules": {
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "assert": "either"
      }
    ],
    "react/jsx-props-no-spreading": "off",
    "react/no-unknown-property": [2, { "ignore": ["class"] }]
  },
  "settings": {
    "react": {
      "pragma": "h"
    }
  }
}
