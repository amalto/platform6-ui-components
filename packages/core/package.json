{
  "name": "@platform6/components",
  "version": "0.0.1",
  "description": "Platform 6 UI components",
  "repository": {
    "type": "git",
    "url": "https://github.com/amalto/platform6-ui-components"
  },
  "license": "MIT",
  "main": "dist/index.cjs.js",
  "module": "dist/custom-elements/index.js",
  "collection": "dist/collection/collection-manifest.json",
  "collection:main": "dist/collection/index.js",
  "es2015": "dist/esm/index.mjs",
  "es2017": "dist/esm/index.mjs",
  "unpkg": "dist/stencil-component/stencil-component.js",
  "types": "dist/types/index.d.ts",
  "files": [
    "dist/",
    "loader/"
  ],
  "scripts": {
    "build": "stencil build --docs",
    "format": "prettier --write ./",
    "postformat": "npm run lint.fix",
    "generate": "stencil generate",
    "lint": "npm run lint.ts && npm run lint.sass",
    "lint.fix": "npm run lint.ts.fix && npm run lint.sass.fix",
    "lint.sass": "stylelint 'src/**/*.scss'",
    "lint.sass.fix": "npm run lint.sass -- --fix",
    "lint.ts": "eslint --cache --ext .js,.jsx,.ts,.tsx ./",
    "lint.ts.fix": "npm run lint.ts -- --fix",
    "start": "stencil build --docs --dev --watch --serve",
    "test": "stencil test --spec --e2e",
    "test.watch": "stencil test --spec --e2e --watchAll",
    "storybook": "start-storybook -p 6006 -s ./dist --ci",
    "build-storybook": "build-storybook"
  },
  "dependencies": {
    "@creativebulma/bulma-tooltip": "^1.2.0",
    "@fortawesome/fontawesome-svg-core": "^1.2.32",
    "@fortawesome/free-brands-svg-icons": "^5.15.1",
    "@fortawesome/free-regular-svg-icons": "^5.15.1",
    "@fortawesome/free-solid-svg-icons": "^5.15.1",
    "@stencil/core": "2.2.0",
    "bulma": "^0.9.1",
    "bulma-calendar": "^6.0.9",
    "choices.js": "9.0.1",
    "lit-html": "^1.3.0"
  },
  "devDependencies": {
    "@babel/core": "^7.12.3",
    "@stencil/eslint-plugin": "^0.3.1",
    "@stencil/postcss": "^1.0.1",
    "@stencil/react-output-target": "0.0.9",
    "@stencil/sass": "^1.3.2",
    "@storybook/addon-a11y": "^6.0.28",
    "@storybook/addon-actions": "^6.0.28",
    "@storybook/addon-essentials": "^6.0.28",
    "@storybook/addon-links": "^6.0.28",
    "@storybook/html": "^6.0.28",
    "@storybook/preset-scss": "^1.0.3",
    "@storybook/theming": "^6.0.28",
    "@storybook/web-components": "^6.0.28",
    "@types/jest": "^26.0.15",
    "@types/js-beautify": "^1.11.0",
    "@typescript-eslint/eslint-plugin": "^4.6.1",
    "@typescript-eslint/parser": "^4.6.1",
    "autoprefixer": "^9.8.6",
    "babel-loader": "^8.1.0",
    "css-loader": "^5.0.0",
    "eslint-plugin-jest": "^24.1.0",
    "jest": "^26.6.3",
    "jest-cli": "^26.6.3",
    "js-beautify": "^1.13.0",
    "prettier": "^2.1.2",
    "puppeteer": "^5.4.1",
    "react-is": "^17.0.1",
    "sass-loader": "^10.0.5",
    "style-loader": "^2.0.0",
    "typescript": "4.0.5"
  }
}
