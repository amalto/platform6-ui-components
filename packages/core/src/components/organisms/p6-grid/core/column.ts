import { Alignment, Direction, SortOrder } from '../../../../shared/types';
import { throwError } from '../../../../utils/error';
import { Column, ColumnDefinition, ColumnId, DataItem } from './entities';

export const MIN_WIDTH = 75;
export const DEFAULT_WIDTH = MIN_WIDTH;
export const INC_WIDTH = 10;

export function defaultGetValue<T extends DataItem>(data: T, context: ColumnDefinition<T>): string {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'getValue' function for ${JSON.stringify(context)}`);
  }

  return `${data[context.field as keyof T]}`;
}

export function defaultSetValue<T extends DataItem>(value: string, data: T, context: ColumnDefinition<T>): T {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'setValue' function for ${JSON.stringify(context)}`);
  }

  return {
    ...data,
    [context.field as keyof T]: value,
  };
}

let columnIds = 0;
export function fromDefinition<T extends DataItem>(column: ColumnDefinition<T>): Column<T> {
  // eslint-disable-next-line no-plusplus
  const id = column.id || column.field || columnIds++;

  return {
    ...column,
    id: `col-${id}`,
    editable: column.editable || false,
    hidden: column.hidden || false,
    width: column.width || 100,
    align: column.align || Alignment.center,
    sortOrder: column.sortOrder || SortOrder.none,
    disableHeaderMenu: column.disableHeaderMenu || false,
    getValue: column.getValue || defaultGetValue,
    setValue: column.setValue || defaultSetValue,
  };
}

export function isHidden<T extends DataItem>(column: Column<T>): boolean {
  return !!column.hidden;
}

export function replaceColumn<T extends DataItem>(updatedColumn: Column<T>, allColumns: Column<T>[]): Column<T>[] {
  return allColumns.map(col => (col.id === updatedColumn.id ? updatedColumn : col));
}

export function move<T extends DataItem>(id: ColumnId, direction: Direction, columns: Column<T>[]): Column<T>[] {
  const from = columns.findIndex(col => col.id === id);
  const to = from + (direction === Direction.left ? -1 : 1);

  if (from < 0 || to < 0 || to >= columns.length) {
    return columns;
  }

  const result = columns.slice();
  const item = result.splice(from, 1);

  return result.slice(0, to).concat(item).concat(result.slice(to, columns.length));
}
