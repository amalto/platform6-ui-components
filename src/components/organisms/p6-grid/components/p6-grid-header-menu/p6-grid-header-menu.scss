:host {
  cursor: context-menu;
  display: flex;
  flex-direction: column;
  background-color: $scheme-main;
  width: 100px;
  padding: $gap-size * 1px;
  font-size: $size-small;
}
