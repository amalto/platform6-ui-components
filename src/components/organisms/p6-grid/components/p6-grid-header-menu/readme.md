# p6-grid-cell

<!-- Auto Generated Below -->

## Properties

| Property                 | Attribute    | Description                   | Type                                                   | Default     |
| ------------------------ | ------------ | ----------------------------- | ------------------------------------------------------ | ----------- |
| `align` _(required)_     | `align`      | Cell alignment                | `Alignment.center \| Alignment.end \| Alignment.start` | `undefined` |
| `column` _(required)_    | --           | the column                    | `Column<Record<string, unknown>>`                      | `undefined` |
| `sortOrder` _(required)_ | `sort-order` | the sort order of this column | `SortOrder.asc \| SortOrder.desc \| SortOrder.none`    | `undefined` |

## Events

| Event            | Description                                 | Type                                                      |
| ---------------- | ------------------------------------------- | --------------------------------------------------------- |
| `p6AlignColumn`  | Update the alignement of the column         | `CustomEvent<ColumnEventDetail<Record<string, unknown>>>` |
| `p6MoveColumn`   | Move the column to the left or to the right | `CustomEvent<MoveColumnDetail<Record<string, unknown>>>`  |
| `p6ResizeColumn` | Resize the column                           | `CustomEvent<ColumnEventDetail<Record<string, unknown>>>` |
| `p6ShowColumn`   | Update the visibility of the column         | `CustomEvent<ColumnEventDetail<Record<string, unknown>>>` |
| `p6SortColumn`   | Sort the column                             | `CustomEvent<ColumnEventDetail<Record<string, unknown>>>` |

## Dependencies

### Used by

- [p6-grid-header-cell](../p6-grid-header-cell)

### Depends on

- [p6-action](../../../../atoms/p6-action)
- [p6-icon](../../../../atoms/p6-icon)

### Graph

```mermaid
graph TD;
  p6-grid-header-menu --> p6-action
  p6-grid-header-menu --> p6-icon
  p6-grid-header-cell --> p6-grid-header-menu
  style p6-grid-header-menu fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
