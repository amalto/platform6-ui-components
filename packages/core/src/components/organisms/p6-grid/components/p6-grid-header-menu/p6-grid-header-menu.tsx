import { Component, ComponentInterface, Event, EventEmitter, h, JSX, Prop } from '@stencil/core';
import { Alignment, Direction, SortOrder } from '../../../../../shared/types';
import { INC_WIDTH } from '../../core/column';
import { AlignColumnDetail, Column, DataItem, MoveColumnDetail, ResizeColumnDetail, ShowColumnDetail, SortColumnDetail } from '../../core/entities';
import { P6GridActionItem } from '../p6-grid-action-item';

@Component({
  tag: 'p6-grid-header-menu',
  styleUrl: 'p6-grid-header-menu.scss',
  shadow: true,
})
export class P6GridHeaderMenu implements ComponentInterface {
  /**
   * the column
   */
  @Prop() column!: Column<DataItem>;

  /**
   * Cell alignment
   */
  @Prop() align!: Alignment;

  /**
   * the sort order of this column
   */
  @Prop() sortOrder!: SortOrder;

  /**
   * Update the alignement of the column
   */
  @Event() p6AlignColumn!: EventEmitter<AlignColumnDetail<DataItem>>;

  /**
   * Move the column to the left or to the right
   */
  @Event() p6MoveColumn!: EventEmitter<MoveColumnDetail<DataItem>>;

  /**
   * Resize the column
   */
  @Event() p6ResizeColumn!: EventEmitter<ResizeColumnDetail<DataItem>>;

  /**
   * Update the visibility of the column
   */
  @Event() p6ShowColumn!: EventEmitter<ShowColumnDetail<DataItem>>;

  /**
   * Sort the column
   */
  @Event() p6SortColumn!: EventEmitter<SortColumnDetail<DataItem>>;

  render(): JSX.Element {
    return (
      <host>
        <div>
          <P6GridActionItem icon="chevron-left" clickHandler={this.moveColumnHandler(Direction.left)} />
          <P6GridActionItem icon="sort" clickHandler={this.sortColumnHandler()} />
          <P6GridActionItem icon="chevron-right" clickHandler={this.moveColumnHandler(Direction.right)} />
        </div>
        <div>
          <P6GridActionItem icon="trash-alt" clickHandler={this.showColumnHandler(false)} />
          <P6GridActionItem icon="minus" clickHandler={this.resizeColumnHandler(INC_WIDTH * -1)} />
          <P6GridActionItem icon="plus" clickHandler={this.resizeColumnHandler(INC_WIDTH)} />
        </div>
        <div>
          <P6GridActionItem icon="align-left" clickHandler={this.alignColumnHandler(Alignment.start)} disabled={this.isDefaultAlignment(Alignment.start)} />
          <P6GridActionItem icon="align-center" clickHandler={this.alignColumnHandler(Alignment.center)} disabled={this.isDefaultAlignment(Alignment.center)} />
          <P6GridActionItem icon="align-right" clickHandler={this.alignColumnHandler(Alignment.end)} disabled={this.isDefaultAlignment(Alignment.end)} />
        </div>
      </host>
    );
  }

  private nextSortOrder(): SortOrder {
    if (this.sortOrder === SortOrder.none) {
      return SortOrder.asc;
    }

    if (this.sortOrder === SortOrder.asc) {
      return SortOrder.desc;
    }

    return SortOrder.none;
  }

  private alignColumnHandler(alignment: Alignment): (event: Event) => void {
    return () => {
      this.p6AlignColumn.emit({ column: { ...this.column, align: alignment } });
    };
  }

  private moveColumnHandler(direction: Direction): (event: Event) => void {
    return () => {
      this.p6MoveColumn.emit({ column: this.column, direction });
    };
  }

  private resizeColumnHandler(offset: number): (event: Event) => void {
    return () => {
      this.p6ResizeColumn.emit({
        column: { ...this.column, width: this.column.width + offset },
      });
    };
  }

  private sortColumnHandler(): (event: Event) => void {
    return () => {
      this.p6SortColumn.emit({
        column: { ...this.column, sortOrder: this.nextSortOrder() },
      });
    };
  }

  private showColumnHandler(visible: boolean): (event: Event) => void {
    return () => {
      this.p6ShowColumn.emit({ column: { ...this.column, hidden: !visible } });
    };
  }

  private isDefaultAlignment(align: Alignment): boolean {
    return align === this.align;
  }
}
