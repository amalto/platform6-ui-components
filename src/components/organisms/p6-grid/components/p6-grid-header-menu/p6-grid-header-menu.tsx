import {
  Component,
  ComponentInterface,
  Event,
  EventEmitter,
  h,
  Prop,
} from "@stencil/core";
import { Alignment, Direction, SortOrder } from "~shared/types";
import { INC_WIDTH } from "../../core/column";
import {
  AlignColumnDetail,
  Column,
  DataItem,
  MoveColumnDetail,
  ResizeColumnDetail,
  ShowColumnDetail,
  SortColumnDetail,
} from "../../core/entities";
import { P6GridActionItem } from "../p6-grid-action-item";

@Component({
  tag: "p6-grid-header-menu",
  styleUrl: "p6-grid-header-menu.scss",
  shadow: true,
})
export class P6GridHeaderMenu implements ComponentInterface {
  /**
   * the column
   */
  @Prop() column!: Column<DataItem>;

  /**
   * Cell alignment
   */
  @Prop() align!: Alignment;

  /**
   * the sort order of this column
   */
  @Prop() sortOrder!: SortOrder;

  /**
   * Update the alignement of the column
   */
  @Event() p6AlignColumn!: EventEmitter<AlignColumnDetail<DataItem>>;

  /**
   * Move the column to the left or to the right
   */
  @Event() p6MoveColumn!: EventEmitter<MoveColumnDetail<DataItem>>;

  /**
   * Resize the column
   */
  @Event() p6ResizeColumn!: EventEmitter<ResizeColumnDetail<DataItem>>;

  /**
   * Update the visibility of the column
   */
  @Event() p6ShowColumn!: EventEmitter<ShowColumnDetail<DataItem>>;

  /**
   * Sort the column
   */
  @Event() p6SortColumn!: EventEmitter<SortColumnDetail<DataItem>>;

  render(): JSX.Element {
    return (
      <host>
        <div>
          <P6GridActionItem
            icon="chevron-left"
            clickHandler={this.moveColumnHandler(Direction.Left)}
          />
          <P6GridActionItem
            icon="sort"
            clickHandler={this.sortColumnHandler()}
          />
          <P6GridActionItem
            icon="chevron-right"
            clickHandler={this.moveColumnHandler(Direction.Right)}
          />
        </div>
        <div>
          <P6GridActionItem
            icon="trash-alt"
            clickHandler={this.showColumnHandler(false)}
          />
          <P6GridActionItem
            icon="minus"
            clickHandler={this.resizeColumnHandler(INC_WIDTH * -1)}
          />
          <P6GridActionItem
            icon="plus"
            clickHandler={this.resizeColumnHandler(INC_WIDTH)}
          />
        </div>
        <div>
          <P6GridActionItem
            icon="align-left"
            clickHandler={this.alignColumnHandler(Alignment.Start)}
            disabled={this.isAlignmentDisabled(Alignment.Start)}
          />
          <P6GridActionItem
            icon="align-center"
            clickHandler={this.alignColumnHandler(Alignment.Center)}
            disabled={this.isAlignmentDisabled(Alignment.Center)}
          />
          <P6GridActionItem
            icon="align-right"
            clickHandler={this.alignColumnHandler(Alignment.End)}
            disabled={this.isAlignmentDisabled(Alignment.End)}
          />
        </div>
      </host>
    );
  }

  private get nextSortOrder(): SortOrder {
    if (this.sortOrder === SortOrder.None) {
      return SortOrder.Asc;
    }

    if (this.sortOrder === SortOrder.Asc) {
      return SortOrder.Desc;
    }

    return SortOrder.None;
  }

  private alignColumnHandler(alignment: Alignment): (event: Event) => void {
    return () => {
      this.p6AlignColumn.emit({ column: { ...this.column, align: alignment } });
    };
  }

  private moveColumnHandler(direction: Direction): (event: Event) => void {
    return () => {
      this.p6MoveColumn.emit({ column: this.column, direction });
    };
  }

  private resizeColumnHandler(offset: number): (event: Event) => void {
    return () => {
      this.p6ResizeColumn.emit({
        column: { ...this.column, width: this.column.width + offset },
      });
    };
  }

  private sortColumnHandler(): (event: Event) => void {
    return () => {
      this.p6SortColumn.emit({
        column: { ...this.column, sortOrder: this.nextSortOrder },
      });
    };
  }

  private showColumnHandler(visible: boolean): (event: Event) => void {
    return () => {
      this.p6ShowColumn.emit({ column: { ...this.column, hidden: !visible } });
    };
  }

  private isAlignmentDisabled(align: Alignment): boolean {
    return align === this.align;
  }
}
