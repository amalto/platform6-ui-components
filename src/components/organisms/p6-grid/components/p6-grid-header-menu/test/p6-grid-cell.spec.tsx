import { newSpecPage } from '@stencil/core/testing';
import { P6GridHeaderMenu } from '../p6-grid-header-menu';

describe('p6-grid-header-menu', () => {
  it('should render an empty cell', async () => {
    const page = await newSpecPage({
      components: [P6GridHeaderMenu],
      html: `<p6-grid-header-menu></p6-grid-header-menu>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
