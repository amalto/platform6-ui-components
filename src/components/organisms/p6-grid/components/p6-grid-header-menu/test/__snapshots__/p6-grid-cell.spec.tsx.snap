// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-grid-header-menu should render an empty cell 1`] = `
<p6-grid-header-menu>
  <mock:shadow-root>
    <host>
      <div>
        <p6-action mode="3">
          <p6-icon name="chevron-left"></p6-icon>
        </p6-action>
        <p6-action mode="3">
          <p6-icon name="sort"></p6-icon>
        </p6-action>
        <p6-action mode="3">
          <p6-icon name="chevron-right"></p6-icon>
        </p6-action>
      </div>
      <div>
        <p6-action mode="3">
          <p6-icon name="trash-alt"></p6-icon>
        </p6-action>
        <p6-action mode="3">
          <p6-icon name="minus"></p6-icon>
        </p6-action>
        <p6-action mode="3">
          <p6-icon name="plus"></p6-icon>
        </p6-action>
      </div>
      <div>
        <p6-action mode="3">
          <p6-icon name="align-left"></p6-icon>
        </p6-action>
        <p6-action mode="3">
          <p6-icon name="align-center"></p6-icon>
        </p6-action>
        <p6-action mode="3">
          <p6-icon name="align-right"></p6-icon>
        </p6-action>
      </div>
    </host>
  </mock:shadow-root>
</p6-grid-header-menu>
`;
